artichoke-backend/src/extn/core/string/trampoline.rs

```diff
@@ -42,31 +44,324 @@ pub fn add(interp: &mut Artichoke, mut value: Value, mut other: Value) -> Result
 
     let mut concatenated = s.clone();
     // XXX: This call doesn't do a check to see if we'll exceed the max allocation
-    //    size and may panic.
+    //    size and may panic or abort.
     concatenated.extend_from_slice(to_append);
     super::String::alloc_value(concatenated, interp)
 }
 
 pub fn push(interp: &mut Artichoke, mut value: Value, mut other: Value) -> Result<Value, Error> {
+    if value.is_frozen(interp) {
+        let s = unsafe { super::String::unbox_from_value(&mut value, interp)? };
+        let message = "can't modify frozen String: "
+            .chars()
+            .chain(s.inspect())
+            .collect::<super::String>();
+        return Err(FrozenError::from(message.into_vec()).into());
+    }
+
     let mut s = unsafe { super::String::unbox_from_value(&mut value, interp)? };
+    if let Ok(int) = other.try_convert_into::<i64>(interp) {
+        return match s.encoding() {
+            Encoding::Utf8 => {
+                // Safety:
+                //
+                // The string is reboxed before any intervening operations on the
+                // interpreter.
+                // The string is reboxed without any intervening mruby allocations.
+                unsafe {
+                    let string_mut = s.as_inner_mut();
+                    // XXX: This call doesn't do a check to see if we'll exceed the max allocation
+                    //    size and may panic or abort.
+                    string_mut
+                        .try_push_codepoint(int)
```

Comment:
Rust is crazy cool with it's conversions here. I was just playing with `try_push_codepoint` expecting it to not work, but of course it does work. I will play with this one day, since `char.try_from(192)` giving unicode `\u00C0` is very cool.

Reply:
🦀 🚀